### ISSUE TITLE
[data grid] Autocomplete edit input cell

### ISSUE THREAD

**Issue by stamahto** [mui/mui-x#4813]
### Summary 💡

Should work like normal Autocomplete component. Unlike 'singleSelect', users can type in input and search options quickly.

### Examples 🌈

![image](https://user-images.githubusercontent.com/57343418/167417216-aa99f655-ac67-4777-a541-1399b5a90b41.png)

### Motivation 🔦

Autocomplete provides more freedom for users and also developers than select (singleSelect), like: freeSolo, getOptionLabel or multiple props.

### Order ID 💳 (optional)

#31956

**Comment by alexfauquette** [comment]
I see multiple points in this proposal:

- some DX improvement, like `getOptionLabel`
- some UX with the ability to search in long list of options
- new column type with the ability to select multiple items (see #4410 for this point)

**Comment by stamahto** [comment]
> I see multiple points in this proposal:
> 
> * some DX improvement, like `getOptionLabel`
> * some UX with the ability to search in long list of options
> * new column type with the ability to select multiple items (see [[data grid] Add `multipleSelect` column type #4410](https://github.com/mui/mui-x/issues/4410) for this point)

Yes, you're right. Thank you 👍 
The best way is not create column type 'autocomplete', but add functionality to existing ones (i.e. `getOptionLabel` can be used for any column type), just add `multipleSelect`.

What I haven't mentioned yet is that if I don't define `valueOptions` for column type: single/multiSelect, I would expect default options as grouped/unique data from the current column.
This is very common in my practise, also **mui-datatables** do that for filter type dropdown
![image](https://user-images.githubusercontent.com/57343418/167686649-c37dd5da-e418-4373-9e61-815b8fb8cf69.png)

**Comment by alexfauquette** [comment]
Yes, `getOptionLabel` could be used for both `singleSelect` and a `multipleSelect` columns. For other types I don't think so. You rarely expect to see a list of options when entering a date or a number

I'm not sure to fully understand your last point `valueOptions` is here to define a finite set of options the user can select. If you do not provide a `valueOptions` parameters, it is similar to using `string` column type. 
I assume your expectation is the API allows you to easily get the array of unique data present in the grid such as you can make them available in an autocomplete. 

But that would concern only the filtering aspect

**Comment by stamahto** [comment]
I believe that in many apps, you always have some information according to which you want to group data or just get the most data consistency (including case sensitivity) for something. Imagine, for example, a warehouse where individual items have a manufacturer, type, etc.
![image](https://user-images.githubusercontent.com/57343418/168149569-80f0615a-431d-4078-a467-4f5255918e41.png)
Currently, to get these values, I have to manually write for each column something like: 
`const manufacturers: string[] = useMemo(() => uniq(data.map(m => m.manufacturer)), [data]);`

I think this is a much better way to automatically display options than as an empty field, **as long as the** `valueOptions` **are not defined** in the column.

I really appreciate your time, thank you.

**Comment by alexfauquette** [comment]
Ok, I better understand the use case.

The technical challenge is to provide access throw the API to the list of unique values. Once done, this list of values could be reused wherever we want: a new custom `renderEditCell` or a `filterInput`, ...

**Comment by adamrneary** [comment]
@alexfauquette we have recently converted to "Pro" and have been tinkering with this problem. I think until this receives the necessary upvotes to get a canonical solution, perhaps a working example in the docs would get us there. I've managed to get us this far:

```typescript
function AutocompleteEditCell<
  T extends { value: string; label: string },
  Multiple extends boolean = false,
  DisableClearable extends boolean = false,
  FreeSolo extends boolean = false
>({
  id,
  value,
  field,
  options,
  disableClearable,
  multiple,
  freeSolo,
}: GridRenderEditCellParams & {
  options: UseAutocompleteProps<
    T,
    Multiple,
    DisableClearable,
    FreeSolo
  >['options'];
  disableClearable: DisableClearable;
  multiple: Multiple;
  freeSolo: FreeSolo;
}) {
  const apiRef = useGridApiContext();
  const handleValueChange = (
    _: any,
    newValue: AutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) => {
    apiRef.current.setEditCellValue({
      id,
      field,
      // @ts-expect-error i can't figure out how to use AutocompleteValue
      value: typeof newValue === 'string' ? value : newValue?.value || '',
    });
  };

  return (
    <Autocomplete<T, Multiple, DisableClearable, FreeSolo>
      fullWidth
      disableClearable={disableClearable}
      multiple={multiple}
      options={options}
      freeSolo={freeSolo}
      // @ts-expect-error i can't figure out how to use AutocompleteValue
      value={options.find((o) => o.value === value)?.label || ''}
      onChange={handleValueChange}
      renderInput={(params) => <TextField {...params} />}
    />
  );
}
```

and this is referenced thusly:

```typescript
{
    field: 'naicsCode',
    headerName: 'Industry',
    width: 250,
    editable: true,
    valueFormatter: ({ value }) => NAICS_OBJECT[value] ?? ''
    renderEditCell: (params) => (
      <AutocompleteEditCell
        {...params}
        options={NAICS_OPTIONS}
        freeSolo={false}
        multiple={false}
        disableClearable
      />
    ),  
},
```

First, if you see anything obvious we could be doing to clear up those two `@ts-expect-error`s I'd love your eyes on those! But more broadly for this task, it seems like something to this effect could go into the docs to explain how to do it manually? Similar to [https://mui.com/x/react-data-grid/editing/#usage-with-mui-x-date-pickers](https://mui.com/x/react-data-grid/editing/#usage-with-mui-x-date-pickers) perhaps?

**Comment by stamahto** [comment]
Thanks @adamrneary for example. It's been a while so I have written solution too. Here is my example: 
```
import { Autocomplete, TextField } from "@mui/material";
import { GridRenderEditCellParams, useGridApiContext } from "@mui/x-data-grid-pro";

interface AutocompleteEditInputCellProps {
    params: GridRenderEditCellParams,
    options: any[] | undefined,
    freeSolo?: boolean,
    multiple?: boolean,
    getOptionLabel?: (option: any) => string
}

export function AutocompleteEditInputCell(props: AutocompleteEditInputCellProps) {
    const apiRef = useGridApiContext();

    const handleChange = (event: React.SyntheticEvent<Element, Event>, newValue: any) => {
        if (props.multiple)
            event.stopPropagation();

        apiRef.current.setEditCellValue({ id: props.params.id, field: props.params.field, value: newValue });
    };

    const getValue = () => {
        if (props.params.value)
            return props.params.value;

        if (props.multiple)
            return [];

        return null;
    }

    return (
        <Autocomplete
            value={getValue()}
            onChange={handleChange}
            onInputChange={(event, value) => (props.freeSolo && !props.multiple && event) && handleChange(event, value)}
            fullWidth
            multiple={props.multiple}
            options={props.options ?? []}
            freeSolo={props.freeSolo}
            autoHighlight
            getOptionLabel={props.getOptionLabel}
            renderInput={(params) => <TextField {...params} error={props.params.error} />}
        />
    );
}
```

In function `handleChange` needs to be `event.stopPropagation();` for multiple prop, because if you hit enter, you don't want to exit edit mode in Grid.

**Comment by Anand24816** [comment]
Yes, I am using Autocomplete in 4 columns in a grid, since it is far better for large drop down lists since it allows user to search from the list by entering first few characters. Though I need single option select only.
My Upvote.

On the other note:-
I have done custom component setup as suggested in docs and similar to what @stamhato has done, and it has been working like charm till  "@mui/x-data-grid-pro": "5.16.0",

When I upgrade my data-grid-pro version to `5.17.1` it starts malfunctioning, the moment the focus goes in the autocomplete cell, my `processRowUpdate` get triggered and the row got saved. 
To save from this bug, I have reverted back to 5.16.0.

I am not sure if I should be logging a new issue for this or if this comment of mine will catch the attention of the material-ui team.

**Comment by alexfauquette** [comment]
@adamrneary You should open another issue with a minimal reproduction of your customization that bugs in v5.17.1

Otherwise, this bug/regression will be lost into other comments asking for feature

**Comment by oliviertassinari** [comment]
I think that it could be great to replace all uses of `Select` for an `Autocomplete`, I think that it would yield a better UX for end-users. e.g. for the `singleSelect` column type edit more.

- Notion (singleSelect): 
<img width="337" alt="Screenshot 2022-09-09 at 17 30 45" src="https://user-images.githubusercontent.com/3165635/189387025-e362623b-e8c1-4101-8ced-2f177df52254.png">

- Airtable (singleSelect):
<img width="199" alt="Screenshot 2022-09-09 at 17 42 30" src="https://user-images.githubusercontent.com/3165635/189389333-8161c0df-559d-42ea-a38a-fc881a7483e7.png">

- Google spreadsheet
<img width="234" alt="Screenshot 2022-09-09 at 17 44 21" src="https://user-images.githubusercontent.com/3165635/189389632-1f0eb9c0-3fef-4e72-b09c-bbef8e8f38a1.png">

**Comment by oliviertassinari** [comment]
I think that it could be great to replace all uses of `Select` for an `Autocomplete`, I think that it would yield a better UX for end-users. e.g. for the `singleSelect` column type edit more.

- Notion (singleSelect): 
<img width="337" alt="Screenshot 2022-09-09 at 17 30 45" src="https://user-images.githubusercontent.com/3165635/189387025-e362623b-e8c1-4101-8ced-2f177df52254.png">

- Airtable (singleSelect):
<img width="199" alt="Screenshot 2022-09-09 at 17 42 30" src="https://user-images.githubusercontent.com/3165635/189389333-8161c0df-559d-42ea-a38a-fc881a7483e7.png">

- Google spreadsheet
<img width="234" alt="Screenshot 2022-09-09 at 17 44 21" src="https://user-images.githubusercontent.com/3165635/189389632-1f0eb9c0-3fef-4e72-b09c-bbef8e8f38a1.png">

**Comment by stamahto** [comment]
> I agree with the UX recommendation made by the author of this issue. I think that it could be great to replace **most** uses of `Select` for an `Autocomplete`, I assume that it would yield a better UX for end-users.
> 
> IMHO, the only question is: Should we always use the `Autocomplete` over the `Select`? Airtable said yes. Notion said no: use a Select if there are always fewer than 5 options.

Well, I don't know the answer either, but here are some of my thoughts:

- Does the user expect to be able to search by typing? Or what can be an advantage of not being able to write to the input.
- What would the types for a column look like? (`singleSelect`, `multiSelect`, `singleAutocomplete`, `multiAutocomplete`)
- Will it affect the render speed?

**Comment by AXLShorts** [comment]
I know its a lot to ask but can anyone provide me a snippet of a working autocomplete functionality inside a cell in the datagrid. I am unable to implement it

**Comment by stamahto** [comment]
> I know its a lot to ask but can anyone provide me a snippet of a working autocomplete functionality inside a cell in the datagrid. I am unable to implement it

My code few comments up here is updated and works pretty well: https://github.com/mui/mui-x/issues/4813#issuecomment-1190009699
The usage is: `renderEditCell: params => <AutocompleteEditInputCell params={params} options={["A", "B"]} />`

**Comment by AXLShorts** [comment]
I got it to work like this:

`import { useState } from "react";
import { DataGrid } from "@mui/x-data-grid";

const UserPage = () => {
  const [users, setUsers] = useState([
    { id: 1, name: "John Doe", email: "john@example.com", role: "basic" },
    { id: 2, name: "Jane Smith", email: "jane@example.com", role: "admin" },
    // Add more user objects as needed
  ]);

  const handleRoleChange = (event, id) => {
    const newUsers = [...users];
    const userIndex = newUsers.findIndex((user) => user.id === id);
    if (userIndex !== -1) {
      newUsers[userIndex].role = event.target.value;
      setUsers(newUsers);
      console.log(newUsers);
    }
  };

  const columns = [
    { field: "name", headerName: "Name", width: 150 },
    { field: "email", headerName: "Email", width: 200 },
    {
      field: "role",
      headerName: "Role",
      width: 120,
      type: "singleSelect",
      valueOptions: ["basic", "admin"],
      editable: true,
    },
  ];

  return (
    <div style={{ height: 400, width: "100%" }}>
      <DataGrid rows={users} columns={columns} />
    </div>
  );
};

export default UserPage;
`


Now it works how you would expect. The cells becomes autocomplete type elements when you start editing them.

**Comment by hcurbelo-hepyco** [comment]
Have a problem when used an Autocomplete inside DataGrid MUI
<img width="1352" alt="Screenshot 2023-06-11 at 19 58 57" src="https://github.com/mui/mui-x/assets/97173659/9768c721-8251-4078-ae19-ee78252bbaa4">

**Comment by AXLShorts** [comment]
I gave a code snippet above. That works

**Comment by hcurbelo-hepyco** [comment]
@AXLShorts Thk, this issue is solved

**Comment by arvindkannan** [comment]
> Have a problem when used an Autocomplete inside DataGrid MUI
> 
> <img width="1352" alt="Screenshot 2023-06-11 at 19 58 57" src="https://github.com/mui/mui-x/assets/97173659/9768c721-8251-4078-ae19-ee78252bbaa4">
> 
> 

I also have same issue where in system displays [object Object] using autocomplete inside mui data grid edit mode

**Comment by aeftimia** [comment]
For what it's worth, setting `renderCell` and `renderEditCell` to the autocomplete box with `editable` seems to work fine for the community version for me.

**Comment by sureshvv** [comment]
> I got it to work like this:
> 
> ```js
> import { useState } from "react";
> import { DataGrid } from "@mui/x-data-grid";
> 
> const UserPage = () => {
>   const [users, setUsers] = useState([
>     { id: 1, name: "John Doe", email: "john@example.com", role: "basic" },
>     { id: 2, name: "Jane Smith", email: "jane@example.com", role: "admin" },
>     // Add more user objects as needed
>   ]);
> 
>   const handleRoleChange = (event, id) => {
>     const newUsers = [...users];
>     const userIndex = newUsers.findIndex((user) => user.id === id);
>     if (userIndex !== -1) {
>       newUsers[userIndex].role = event.target.value;
>       setUsers(newUsers);
>       console.log(newUsers);
>     }
>   };
> 
>   const columns = [
>     { field: "name", headerName: "Name", width: 150 },
>     { field: "email", headerName: "Email", width: 200 },
>     {
>       field: "role",
>       headerName: "Role",
>       width: 120,
>       type: "singleSelect",
>       valueOptions: ["basic", "admin"],
>       editable: true,
>     },
>   ];
> 
>   return (
>     <div style={{ height: 400, width: "100%" }}>
>       <DataGrid rows={users} columns={columns} />
>     </div>
>   );
> };
> 
> export default UserPage;
> ```
> 
> Now it works how you would expect. The cells becomesg  autocomplete type elements when you start editing them. But now I'm having slight trouble in handling the change. Any idea on how to track change here?

who is calling handleRoleChange and how does it know to di it?
